### Embeddings

Convert Text into Vectors

In [1]:
import os
from dotenv import load_dotenv
load_dotenv()

True

In [2]:
os.environ["OPENAI_API_KEY"]=os.getenv("OPENAI_API_KEY")

In [3]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(
    model="text-embedding-3-large",
    # With the `text-embedding-3` class
    # of models, you can specify the size
    # of the embeddings you want returned.
    # dimensions=1024
)

In [4]:
embeddings

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x11aba92b0>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x11aba9be0>, model='text-embedding-3-large', dimensions=None, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [5]:
text="This is a tutoriall on Open AI Embeddings"
query_result=embeddings.embed_query(text)
query_result


[-0.004291534423828125,
 0.02911376953125,
 -0.0177001953125,
 -0.045867919921875,
 0.031768798828125,
 -0.0199127197265625,
 0.0145721435546875,
 0.048828125,
 -0.00815582275390625,
 0.01209259033203125,
 -0.01197052001953125,
 0.00872039794921875,
 -0.00463104248046875,
 -0.02166748046875,
 0.005039215087890625,
 0.006649017333984375,
 0.0125274658203125,
 0.03717041015625,
 -0.02471923828125,
 -0.0203857421875,
 0.0158233642578125,
 0.022064208984375,
 -0.048614501953125,
 -0.0390625,
 0.02191162109375,
 -0.0251617431640625,
 0.01525115966796875,
 0.040740966796875,
 -0.00791168212890625,
 0.04278564453125,
 0.0079193115234375,
 -0.020263671875,
 0.00995635986328125,
 -0.003978729248046875,
 -0.0169219970703125,
 0.01433563232421875,
 0.054962158203125,
 0.034698486328125,
 -0.05572509765625,
 0.033447265625,
 0.0242919921875,
 0.00795745849609375,
 0.0017290115356445312,
 -0.01346588134765625,
 -0.04107666015625,
 -0.050933837890625,
 0.00418853759765625,
 -0.01776123046875,
 -0.01

In [6]:
len(query_result)

3072

Irrespective of how long the sentence is, Open AI embedding will convert it into 3072 dimensions.

In [14]:
from langchain_community.document_loaders import TextLoader
loader=TextLoader('/Users/sanjuktabaruah/Desktop/coursera/Krishnaink/agenticAI/2-Langchain Basics/2.1-DataIngestion/speech.txt')
docs=loader.load()

In [15]:
from langchain_text_splitters import RecursiveCharacterTextSplitter
text_splitter=RecursiveCharacterTextSplitter(chunk_size=500,chunk_overlap=50)
final_documents=text_splitter.split_documents(docs)
final_documents

[Document(metadata={'source': '/Users/sanjuktabaruah/Desktop/coursera/Krishnaink/agenticAI/2-Langchain Basics/2.1-DataIngestion/speech.txt'}, page_content='rtificial intelligence (AI) is transforming the world in ways that were unimaginable just a decade ago. From self-driving cars to intelligent personal assistants, AI technologies are becoming an integral part of daily life. Businesses are leveraging AI to optimize operations, reduce costs, and create new products and services. At the same time, governments and organizations are exploring regulations to ensure AI is developed and deployed responsibly, addressing concerns such as privacy, bias,'),
 Document(metadata={'source': '/Users/sanjuktabaruah/Desktop/coursera/Krishnaink/agenticAI/2-Langchain Basics/2.1-DataIngestion/speech.txt'}, page_content='addressing concerns such as privacy, bias, and accountability.'),
 Document(metadata={'source': '/Users/sanjuktabaruah/Desktop/coursera/Krishnaink/agenticAI/2-Langchain Basics/2.1-DataIng

In [ ]:
final_documents[0]

'rtificial intelligence (AI) is transforming the world in ways that were unimaginable just a decade ago. From self-driving cars to intelligent personal assistants, AI technologies are becoming an integral part of daily life. Businesses are leveraging AI to optimize operations, reduce costs, and create new products and services. At the same time, governments and organizations are exploring regulations to ensure AI is developed and deployed responsibly, addressing concerns such as privacy, bias,'

In [19]:
embeddings.embed_query(final_documents[0].page_content)

[-0.020490441471338272,
 0.011126982048153877,
 -0.014052188955247402,
 -0.030091838911175728,
 0.014612036757171154,
 0.029224073514342308,
 0.008096803911030293,
 -0.021736104041337967,
 0.0008082807762548327,
 0.015017926692962646,
 -0.011658838018774986,
 -0.030483731999993324,
 0.014486070722341537,
 0.01045516412705183,
 -0.0333949439227581,
 -0.006938618142157793,
 0.03347891941666603,
 -0.010385183617472649,
 -0.011518876068294048,
 -0.014514063484966755,
 -0.004422800615429878,
 0.01034319493919611,
 -0.0502183772623539,
 -0.006494238507002592,
 0.002010204829275608,
 0.007669919636100531,
 0.020000575110316277,
 0.008950572460889816,
 -0.06958912312984467,
 0.013177425600588322,
 0.02481526881456375,
 -0.003572531044483185,
 0.039161376655101776,
 -0.040337059646844864,
 0.0017556487582623959,
 0.03429069742560387,
 0.011917767114937305,
 0.007459976710379124,
 0.0076419273391366005,
 0.01553578581660986,
 0.04431198164820671,
 -0.014612036757171154,
 0.005518003366887569,
 0